stl/inc/xstring

```diff
@@ -3980,6 +3980,25 @@ public:
         }
     }
 
+#if _HAS_CXX23
+    template <class _Operation>
+    constexpr void resize_and_overwrite(_CRT_GUARDOVERFLOW const size_type _Newsize, _Operation _Op) {
+        if (_Mypair._Myval2._Myres < _Newsize) {
+            _Reallocate_grow_by(
+                _Newsize - size(), [](_Elem* const _New_ptr, const _Elem* const _Old_ptr, const size_type _Oldsize) {
+                    _Traits::copy(_New_ptr, _Old_ptr, _Oldsize + 1);
+                });
+        }
+
+        const auto _Resultsize = _STD move(_Op)(data(), _Newsize);
+#if _CONTAINER_DEBUG_LEVEL > 0
+        _STL_VERIFY(_Resultsize >= 0, "the returned size can't be smaller than 0");
+        _STL_VERIFY(_Resultsize <= _Newsize, "the returned size can't be greater than passed size");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        _Eos(_Resultsize);

```

Comment:
When we need to reallocate, `_Reallocate_grow_by()` orphans all iterators, which is good:
https://github.com/microsoft/STL/blob/178b8406a5647507e759442960f17cd33045cdf5/stl/inc/xstring#L4566
However, when we can reuse the existing capacity, `_Eos()` just updates the `size()` and scribbles a null terminator:
https://github.com/microsoft/STL/blob/178b8406a5647507e759442960f17cd33045cdf5/stl/inc/xstring#L4598-L4600

Thus, I believe that when it doesn't reallocate, this implementation leaves all iterators valid:
* Iterators to the first part of the string overwritten by `_Op`
* Iterators to the second part of the string, within `n` (aka `_Newsize`) but beyond the size that `_Op` wrote into and returned
* Iterators to the third part of the string, within its original size but beyond the passed `n` / `_Newsize`

If I understand WG21-N4901 [string.capacity]/10 "invalidates all pointers and references to the range `[p, p + n]`" correctly, we should always be orphaning all iterators.

Reply:
I don't think we should be orphaning iterators in that case. `[p, p + n]` is a range passed to `op` and it's being invalidated so that the user can't use it after `op` returns. `p` is defined as:
> p be a charT*, such that the range [p, p + n] is valid and this->compare(0, k, p, k) == 0 is true before the call

This indicates that `[p, p + n]` and `*this` can be different ranges, so invaliding one doesn't invalidate the other. Also it would we weird for `resize_and_overwrite` to always invalidate iterators if `resize` does not.

Reply:
Ok, that makes sense, thanks for explaining!

I looked into `string`'s implementation (since I was somewhat confused, being much more familiar with `vector`). I was somewhat surprised to discover that (1) `assign` doesn't invalidate iterators when it overwrites characters, and (2) even though `resize()` apparently doesn't go to any effort to invalidate iterators, when the string size is reduced what happens is that iterators beyond the newly written null terminator *are* detected as invalid:

```
D:\GitHub\STL\out\build\x64>type meow.cpp
#include <iostream>
#include <string>
using namespace std;

int main() {
    string str{"abcdefghijklmnopqrstuvwxyz"};
    auto it = str.begin() + 24;
    cout << *it << endl; // y
    str.assign(26, 'A');
    cout << *it << endl; // A
    str.resize(20);
    cout << static_cast<int>(*it) << endl;
}

D:\GitHub\STL\out\build\x64>cl /EHsc /nologo /W4 /MTd meow.cpp && meow
meow.cpp
y
A
[***DIALOG***]
Debug Assertion Failed!
Expression: cannot dereference string iterator because it is out of range (e.g. an end iterator)
```

This is because, although `_Eos` really does just update the size and scribble the null terminator, the iterator debugging logic has each iterator check its parent string's size, failing if the iterator points beyond the current size. I forgot about that part.

Thus, `resize_and_overwrite` will correctly invalidate iterators beyond the current null terminator, so your code is correct as written. Yay! I appreciate your double-checking of my occasionally bogus advice. :tada: :heart_eyes_cat: :joy_cat: